lib/logstash/outputs/elasticsearch/data_stream_support.rb

```diff
@@ -50,48 +49,59 @@ def data_stream_name(event)
-          raise LogStash::ConfigurationError, "Ambiguous configuration, please remove data stream specific settings: #{data_stream_params.keys}"
-        end
+        check_disabled_data_stream_config!(params)
         return false
       when true
-        if invalid_data_stream_params.any?
-          @logger.error "Invalid data stream configuration, following parameters are not supported:", invalid_data_stream_params
-          raise LogStash::ConfigurationError, "Invalid data stream configuration: #{invalid_data_stream_params.keys}"
-        end
-        if ecs_compatibility == :disabled
-          if ::Gem::Version.create(LOGSTASH_VERSION) < ::Gem::Version.create(DATA_STREAMS_REQUIRES_ECS_LS_VERSION)
-            @deprecation_logger.deprecated "In a future release of Logstash, the Elasticsearch output plugin's `data_stream => true` will require the plugin to be run in ECS compatibility mode. " + ENABLING_ECS_GUIDANCE
-          else
-            @logger.error "Invalid data stream configuration; `ecs_compatibility` must not be `disabled`. " + ENABLING_ECS_GUIDANCE
-            raise LogStash::ConfigurationError, "Invalid data stream configuration: `ecs_compatibility => disabled`"
-          end
-        end
+        check_enabled_data_stream_config!(params)
         return true
-      else
-        use_data_stream = data_stream_default(data_stream_params, invalid_data_stream_params)
-        if use_data_stream
-          @logger.info("Config is compliant with data streams. `data_stream => auto` resolved to `true`")
-        elsif data_stream_params.any?
-          # DS (auto) disabled but there's still some data-stream parameters (and no `data_stream => false`)
-          @logger.error "Ambiguous configuration; data stream settings are present, but data streams are not enabled", data_stream_params
-          raise LogStash::ConfigurationError, "Ambiguous configuration, please set data_stream => true " +
-              "or remove data stream specific settings: #{data_stream_params.keys}"
+      else # data_stream => auto or not set
+        use_data_stream = data_stream_default(params)
+
+        check_disabled_data_stream_config!(params) unless use_data_stream
+
+        @logger.info("Data streams auto configuration (`data_stream => auto` or unset) resolved to `#{use_data_stream}`")
+        return use_data_stream
+      end
+    end
+
+    def check_enabled_data_stream_config!(params)
+      invalid_data_stream_params = invalid_data_stream_params(params)
+
+      if invalid_data_stream_params.any?
+        @logger.error "Invalid data stream configuration, the following parameters are not supported:", invalid_data_stream_params
+        raise LogStash::ConfigurationError, "Invalid data stream configuration: #{invalid_data_stream_params.keys}"
+      end
+
+      if ecs_compatibility == :disabled
+        if ::Gem::Version.create(LOGSTASH_VERSION) < ::Gem::Version.create(DATA_STREAMS_REQUIRES_ECS_LS_VERSION)
```

Comment:
Since we are moving to a more assertion like checks, like:
```ruby
if ! condition
  raise Error
end
if ! another_condition
  raise AnotherError
end  
```

maybe would be best to invert the `if` condition and "cut" the `else`:

```ruby
if ::Gem::Version.create(LOGSTASH_VERSION) >= ::Gem::Version.create(DATA_STREAMS_REQUIRES_ECS_LS_VERSION)
  @logger.error "Invalid data stream configuration; `ecs_compatibility` must not be `disabled`. " + ENABLING_ECS_GUIDANCE
  raise LogStash::ConfigurationError, "Invalid data stream configuration: `ecs_compatibility => disabled`"
end

@deprecation_logger.deprecated "In a future release of Logstash, the Elasticsearch output plugin's `data_stream => true` will require the plugin to be run in ECS compatibility mode. " + ENABLING_ECS_GUIDANCE
```

Reply:
`DATA_STREAMS_ENABLED_BY_DEFAULT_LS_VERSION` and `DATA_STREAMS_REQUIRES_ECS_LS_VERSION` both ECS and Datastream on by default was introduced with version `8.0.0` so maybe could collapsed in one constant and 

```
::Gem::Version.create(LOGSTASH_VERSION) >= ::Gem::Version.create(DATA_STREAMS_REQUIRES_ECS_LS_VERSION)
```

```
ds_default = ::Gem::Version.create(LOGSTASH_VERSION) >= ::Gem::Version.create(DATA_STREAMS_ENABLED_BY_DEFAULT_LS_VERSION)
```

could be extracted in a meaningful method. WDYT?

Reply:
Yes, it sounds good to me to keep only one constant and extract this logic into a meaningful method! I tried to not touch the existing code logic to keep the refactoring as small as possible. There's also this [duplicate logic](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/main/lib/logstash/outputs/elasticsearch/data_stream_support.rb#L233) `validate_dataset_identifier` & `validate_namespace_identifier` that could be collapsed into one.

If you think it would be okay to address those changes on this PR, I'd be glad to do so.

Reply:
Maybe it's better to do these other changes in a follow PR, to keep the scope of this close to the original problem.